lib/ProductOpener/Ecoscore.pm

```diff
@@ -0,0 +1,240 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2020 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::Ecoscore - compute the Ecoscore environemental grade of a food product
+
+=head1 SYNOPSIS
+
+C<ProductOpener::Ecoscore> is used to compute the Ecoscore environmental grade
+of a food product.
+
+=head1 DESCRIPTION
+
+The modules implements the Eco-Score computation as defined by a collective that Open Food Facts is part of.
+
+It is based on the French AgriBalyse V3 database that contains environmental impact values for 2500 food product categories.
+
+AgriBalyse provides Life Cycle Analysis (LCA) values for food products categories,
+and some adjustments to the score are made for actual specific products using data about labels, origins of ingredients, packagings etc.
+
+=cut
+
+package ProductOpener::Ecoscore;
+
+use utf8;
+use Modern::Perl '2017';
+use Exporter    qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN
+{
+	use vars       qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+
+		&load_agribalyse_data
+		&compute_ecoscore
+
+		);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK ;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Tags qw/:all/;
+
+use Text::CSV();
+
+my %agribalyse = ();
+
+=head1 FUNCTIONS
+
+=head2 load_agribalyse_data( $product_ref )
+
+Loads the AgriBalyse database.
+
+=cut
+
+sub load_agribalyse_data() {
+
+	my $agribalyse_details_by_step_csv_file = $data_root . "/ecoscore/agribalyse/AGRIBALYSE3.0.1_vf.csv.2";
+	
+	my $headers_ref;
+	my $rows_ref = [];
+
+	my $encoding = "UTF-8";
+
+	$log->debug("opening agribalyse CSV file", { file => $agribalyse_details_by_step_csv_file }) if $log->is_debug();
+
+	my $csv_options_ref = { binary => 1, sep_char => "," };    # should set binary attribute.
+
+	my $csv = Text::CSV->new ( $csv_options_ref )
+		or die("Cannot use CSV: " . Text::CSV->error_diag ());
+
+	if (open (my $io, "<:encoding($encoding)", $agribalyse_details_by_step_csv_file)) {
+
+		# @$headers_ref = $csv->header ($io, { detect_bom => 1 });
+		# the header function crashes with some csv files... use getline instead
```

Comment:
Is this related to the `Strings with code points over 0xFF may not be mapped into in-memory file handles` error? I worked around that one in `update_packager_codes.pl` by setting the encoding 
only after the headers were read.

https://github.com/openfoodfacts/openfoodfacts-server/blob/12d17130cb5cca3db85a70cbcfc3a5b6fd5ab5b1/scripts/update_packager_codes.pl#L209

...

https://github.com/openfoodfacts/openfoodfacts-server/blob/12d17130cb5cca3db85a70cbcfc3a5b6fd5ab5b1/scripts/update_packager_codes.pl#L219-L234

Reply:
I don't remember if there was an error message, I think it was a segmentation fault.

In EcoScore.pm we don't use the header line at all in fact, so I'm going to just remove the comment.
I copied it from Producers.pm (whe we do need the headers line):

````
			# @$headers_ref = $csv->header ($io, { detect_bom => 1 });
			# the header function crashes with some csv files... use getline instead
			my $row_ref;

			while ((not defined $row_ref) and ($row_ref = $csv->getline ($io))) {
			}

			if (defined $row_ref) {

				@{$headers_ref} = @{$row_ref};

				while ($row_ref = $csv->getline ($io)) {
					push @{$rows_ref}, $row_ref;
				}
			}
			else {
				$results_ref->{error} = "Could not read header line in CSV $file: $!";
			}